File: 2.字典序的第K小数字/cpp/440.cpp
class Solution {
public:
    int findKthNumber(int n, int k) {
        auto getCount = [](int prefix, long n) {
            long cur = prefix;
            long next = prefix + 1;
            long count = 0;
            while (cur <= n) {
                count += min(n+1, next) - cur;
                cur *= 10;
                next *= 10;
            }
            return count;
        };
        
        int prefix = 1;
        long count = 1;
        while (count < k) {
            long c = getCount(prefix, n);
            if (count + c > k) {
                prefix *= 10;
                count ++;
            } else {
                prefix++;
                count += c;
            }
        }
        return prefix;
    }
};

File: 02.字典序的第K小数字/go/440.go
//author : baronliu，问题核心在于字典排序里，数字+1，indx增加的量；以及index+1，数字增加的量需要重新换算

func findKthNumber(n int, k int) int {
	p, result := 1, 1
	for p < k {
		c := getCount(result, n)
		if p+c > k {
			//说明锁定范围就在此前缀
			p += 1
			result *= 10
		} else {
			//说明范围不在此前缀，需要进位
			p += c
			result += 1
		}
	}
	return result
}

func getCount(prefix int, n int) int {
	cnt := 0
	for a, b := prefix, prefix+1; a <= n; a, b = a*10, b*10 {
		cnt += min(b, n+1) - a
	}
	return cnt
}

func min(a int, b int) int {
	if a >= b {
		return b
	}
	return a
}